projects/motor_controller_interface/src/motor_controller_fsm.c

```diff
@@ -0,0 +1,322 @@
+    return true;
+  }
+  return false;
+}
+
+static bool prv_cruise_control_forward_guard(const FSM *fsm, const Event *e, void *context) {
+  // We can only go Cruise Control if the direction selector is set to Forward
+  // and the Current is a valid value
+  DriverControlsData data = { 0 };
+
+  fifo_peek(s_fifo, &data);
+  if (data.brake_state == DRIVER_CONTROLS_BRAKE_DISENGAGED &&
+      data.direction == DRIVER_CONTROLS_FORWARD && data.cruise_control != 0) {
+    // The pop will happen in the output function
+    return true;
+  }
+  return false;
+}
+
+// Motor Controller state definitions
+
+// Note: Cruise Control is invalid in reverse
+FSM_DECLARE_STATE(torque_control_forward);
+FSM_DECLARE_STATE(torque_control_reverse);
+FSM_DECLARE_STATE(cruise_control_forward);
+FSM_DECLARE_STATE(mechanical_braking);
+
+// Motor Controller transition table definitions
+
+FSM_STATE_TRANSITION(torque_control_forward) {
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_torque_control_forward_guard, torque_control_forward);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_torque_control_reverse_guard, torque_control_reverse);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_cruise_control_forward_guard, cruise_control_forward);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO, prv_mechanical_braking_guard,
+                             mechanical_braking);
+}
+
+FSM_STATE_TRANSITION(torque_control_reverse) {
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_torque_control_forward_guard, torque_control_forward);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_torque_control_reverse_guard, torque_control_reverse);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_cruise_control_forward_guard, cruise_control_forward);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO, prv_mechanical_braking_guard,
+                             mechanical_braking);
+}
+
+FSM_STATE_TRANSITION(cruise_control_forward) {
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_torque_control_forward_guard, torque_control_forward);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_torque_control_reverse_guard, torque_control_reverse);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO,
+                             prv_cruise_control_forward_guard, cruise_control_forward);
+
+  FSM_ADD_GUARDED_TRANSITION(MOTOR_CONTROLLER_INTERFACE_EVENT_FIFO, prv_mechanical_braking_guard,
+                             mechanical_braking);
+}
+
+FSM_STATE_TRANSITION(mechanical_braking) {
```

Comment:
Yeah that's what I was thinking about as well... Ideally that'd never happen, but if it does, nothing currently guards against that.

Not sure what the best way to model that in our state machine implementation.

The only thing that I can think of atm is to validate the message in the FIFO in each guard function for each possible case, and then discard if the message is invalid?

Or maybe we could validate each message before pushing into the FIFO for each of the message types we accept..

I guess if we had a default state transition as a "catch-all" case, we could force that transition, which would discard the FIFO message in the output function? I think the problem with that is we would have to keep track of whether we transitioned (or the previous state), since we would need to transition back to the state that entered the default state.

I'm not really a fan of validating each message, because I think that's prone to error, but I'm not sold either way.

Thoughts?